pkg/controller/cloud/node_controller.go

```diff
@@ -55,6 +57,12 @@ type CloudNodeController struct {
 	cloud cloudprovider.Interface
 
 	nodeStatusUpdateFrequency time.Duration
+
+	// Set of node names that have been initialized recently.
+	// Used to discard UpdateCloudNode events in which the received node object still
+	// has the cloud taint, even though the node has actually been initialized.
+	recentlyInitializedNodes      sets.String
```

Comment:
Still wrapping my head around this PR so sorry in advance. I'm not entirely happy with the semantics around "recently initialized" 🤔 IMHO, it's a bit difficult to grok how this is used to ensure we're not initializing a node that is already being initialized, especially with the way it's used differently between OnAdd and OnUpdate events. Thoughts on this @cheftako?

Reply:
Maybe I still haven't managed to explain properly. Does a picture with a timeline help? This is what (I think) is currently happening:

![node-init](https://user-images.githubusercontent.com/11710676/55387572-f0030900-5531-11e9-9b84-e3b26cb5fd82.png)
With the fix in this PR, all the update events are ignored because the node name is remembered from t=35 (after the initialization is first completed) so that the update events from before t=35 can be ignored. In this case the Update events can all be ignored and be cleared out from the event queue very quickly, allowing other nodes to be initialized. For example using the fix, in [these logs](https://gist.github.com/tghartland/1bdaffc718984d8f0faa1a02bdb53c13) there are many events for minion-41 that are only ignored because the name is remembered, which would otherwise have been re-processed.

When the update events are processed the actual Add event initialization was done in the recent past, not both happening at the same time, which is why I chose the name. If there is a shorter name that works I would change to that, since recentlyInitializedNodesMutex is a bit long.

I hope this makes things clearer.